drivers/pwm/pwm_rpi_pico.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2022, Joep Buruma
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT raspberrypi_pico_pwm
+
+#include <device.h>
+#include <drivers/pwm.h>
+#include <drivers/pinctrl.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pwm_rpi_pico, CONFIG_PWM_LOG_LEVEL);
+
+/* pico-sdk includes */
+#include <hardware/pwm.h>
+#include <hardware/structs/pwm.h>
+
+#define PWM_RPI_PICO_COUNTER_TOP_MAX    UINT16_MAX
+#define PWM_RPI_NUM_CHANNELS            (16U)
+
+struct pwm_rpi_slice_config {
+	uint8_t integral;
+	uint8_t frac;
+	bool phase_correct;
+};
+
+struct pwm_rpi_config {
+	/*
+	 * pwm_controller is the start address of the pwm peripheral.
+	 */
+	pwm_hw_t *pwm_controller;
+	struct pwm_rpi_slice_config slice_configs[NUM_PWM_SLICES];
+	const struct pinctrl_dev_config *pcfg;
+};
+
+static float pwm_rpi_get_clkdiv(const struct device *dev, int slice)
+{
+	const struct pwm_rpi_config *cfg = dev->config;
+
+	/* the divider is a fixed point 8.4 convert to float for use in pico-sdk */
+	return (float)cfg->slice_configs[slice].integral +
+		(float)cfg->slice_configs[slice].integral / 16.0;
+}
+
+static inline int pwm_rpi_channel_to_slice(uint32_t channel)
+{
+	return channel / 2;
+}
+
+static inline int pwm_rpi_channel_to_pico_channel(uint32_t channel)
+{
+	return channel % 2;
+}
```

Comment:
nit seems like you are returning `int` from `uint32_t`, should likely return `uint32_t`, or, cast down if you're sure about the range of values.